File: backend/app/realtime_bus.py
import json
import asyncio
from redis.asyncio import Redis
from app.config import settings
from app.realtime import manager

_channel = "unisync:realtime"
_redis: Redis | None = None


async def _get_redis() -> Redis:
    global _redis
    if _redis is None:
        if not settings.use_redis or not settings.redis_url:
            raise RuntimeError("Redis disabled")
        _redis = Redis.from_url(settings.redis_url)
    return _redis


async def publish_event(user_id: str, payload: dict) -> None:
    if not settings.use_redis or not settings.redis_url:
        return
    redis = await _get_redis()
    await redis.publish(_channel, json.dumps({"user_id": user_id, "payload": payload}))


async def listen_and_forward(stop_event: asyncio.Event) -> None:
    if not settings.use_redis or not settings.redis_url:
        return
    redis = await _get_redis()
    pubsub = redis.pubsub()
    await pubsub.subscribe(_channel)
    try:
        while not stop_event.is_set():
            message = await pubsub.get_message(ignore_subscribe_messages=True, timeout=1.0)
            if message and message.get("data"):
                data = json.loads(message["data"])
                user_id = data.get("user_id")
                payload = data.get("payload")
                if user_id and payload:
                    await manager.send(user_id, payload)
    finally:
        await pubsub.unsubscribe(_channel)

File: backend2/src/realtimeBus.ts
import Redis from "ioredis";

import { settings } from "./config.js";
import { manager } from "./realtime.js";

const channel = "unisync:realtime";
let redis: Redis | null = null;

function getRedis(): Redis {
  if (!settings.useRedis || !settings.redisUrl) throw new Error("Redis disabled");
  if (!redis) redis = new Redis(settings.redisUrl);
  return redis;
}

export async function publishEvent(userId: string, payload: Record<string, unknown>): Promise<void> {
  if (!settings.useRedis || !settings.redisUrl) return;
  await getRedis().publish(channel, JSON.stringify({ userId, payload }));
}

export async function listenAndForward(signal: AbortSignal): Promise<void> {
  if (!settings.useRedis || !settings.redisUrl) return;
  const subscriber = new Redis(settings.redisUrl);
  await subscriber.subscribe(channel);
  subscriber.on("message", (_channel, message) => {
    try {
      const parsed = JSON.parse(message) as { userId?: string; payload?: Record<string, unknown> };
      if (parsed.userId && parsed.payload) manager.send(parsed.userId, parsed.payload);
    } catch {
      return;
    }
  });
  signal.addEventListener("abort", () => {
    void subscriber.unsubscribe(channel).finally(() => subscriber.quit());
  });
}
